In [6]:
import pandas as pd 
import numpy as np 
import matplotlib.pyplot as plt 
import seaborn as sns 
from scipy import stats 
sns.set()

In [7]:
train = pd.read_csv('C:/Users/nyc15/Desktop/data/train.csv')
test = pd.read_csv('C:/Users/nyc15/Desktop/data/test_x.csv')

In [8]:
y = train.voted.copy()
train = train.drop(columns = 'voted')
raw_data = pd.concat([train,test])

In [9]:
raw_data = raw_data.reset_index()
raw_data.drop(columns = ['level_0','index'],inplace= True)
raw_data= raw_data.reset_index()

In [10]:
raw_data.shape,y.shape

((56915, 77), (45532,))

In [11]:
raw_data = raw_data.drop(index= raw_data[raw_data.familysize >=100].index) # 가족 사이즈 3개 열 (2147483647 , 999 , 100) 제거 

In [12]:
raw_data.shape

(56912, 77)

In [8]:
raw_data[raw_data.familysize >=100]

Empty DataFrame
Columns: [index, QaA, QaE, QbA, QbE, QcA, QcE, QdA, QdE, QeA, QeE, QfA, QfE, QgA, QgE, QhA, QhE, QiA, QiE, QjA, QjE, QkA, QkE, QlA, QlE, QmA, QmE, QnA, QnE, QoA, QoE, QpA, QpE, QqA, QqE, QrA, QrE, QsA, QsE, QtA, QtE, age_group, education, engnat, familysize, gender, hand, married, race, religion, tp01, tp02, tp03, tp04, tp05, tp06, tp07, tp08, tp09, tp10, urban, wf_01, wf_02, wf_03, wr_01, wr_02, wr_03, wr_04, wr_05, wr_06, wr_07, wr_08, wr_09, wr_10, wr_11, wr_12, wr_13]
Index: []

[0 rows x 77 columns]

In [19]:
raw_data.columns

Index(['index', 'QaA', 'QaE', 'QbA', 'QbE', 'QcA', 'QcE', 'QdA', 'QdE', 'QeA',
       'QeE', 'QfA', 'QfE', 'QgA', 'QgE', 'QhA', 'QhE', 'QiA', 'QiE', 'QjA',
       'QjE', 'QkA', 'QkE', 'QlA', 'QlE', 'QmA', 'QmE', 'QnA', 'QnE', 'QoA',
       'QoE', 'QpA', 'QpE', 'QqA', 'QqE', 'QrA', 'QrE', 'QsA', 'QsE', 'QtA',
       'QtE', 'age_group', 'education', 'engnat', 'familysize', 'gender',
       'hand', 'married', 'race', 'religion', 'tp01', 'tp02', 'tp03', 'tp04',
       'tp05', 'tp06', 'tp07', 'tp08', 'tp09', 'tp10', 'urban', 'wf_01',
       'wf_02', 'wf_03', 'wr_01', 'wr_02', 'wr_03', 'wr_04', 'wr_05', 'wr_06',
       'wr_07', 'wr_08', 'wr_09', 'wr_10', 'wr_11', 'wr_12', 'wr_13'],
      dtype='object')

In [15]:
raw_data.age_group.value_counts().index

Index(['10s', '20s', '30s', '40s', '50s', '60s', '+70s'], dtype='object')

In [20]:
raw_data['agenum'] = raw_data['age_group'].map({'10s':10, '20s':20, '30s':30, '40s':40, '50s':50, '60s':60, '+70s':70})

In [54]:
raw_data.corr(method='spearman')[raw_data.corr(method='spearman')['education']>0.2]['education']

education    1.000000
married      0.278788
wr_03        0.220288
agenum       0.507311
Name: education, dtype: float64

## 1=Less than high school, 2=High school, 3=University degree, 4=Graduate degree, 0=무응답

In [10]:
raw_data.education.value_counts()

2    21068
3    19025
4     9817
1     6352
0      650
Name: education, dtype: int64

# education과 상관계수가 높은 데이터 컬럼 몇개 뽑아서 랜덤으로 넣자!
# 랜덤 확률은 컬럼에서 데이터 개수가 많은 게 확률이 높게 지정한다! 
# 연령 0.507311 결혼 유무 0.241386 

In [11]:
raw_data.shape

(56912, 77)

In [12]:
educationdata = raw_data['education'].copy()

In [13]:
(educationdata == 0).sum()

650

In [14]:
raw_data.columns

Index(['index', 'QaA', 'QaE', 'QbA', 'QbE', 'QcA', 'QcE', 'QdA', 'QdE', 'QeA',
       'QeE', 'QfA', 'QfE', 'QgA', 'QgE', 'QhA', 'QhE', 'QiA', 'QiE', 'QjA',
       'QjE', 'QkA', 'QkE', 'QlA', 'QlE', 'QmA', 'QmE', 'QnA', 'QnE', 'QoA',
       'QoE', 'QpA', 'QpE', 'QqA', 'QqE', 'QrA', 'QrE', 'QsA', 'QsE', 'QtA',
       'QtE', 'age_group', 'education', 'engnat', 'familysize', 'gender',
       'hand', 'married', 'race', 'religion', 'tp01', 'tp02', 'tp03', 'tp04',
       'tp05', 'tp06', 'tp07', 'tp08', 'tp09', 'tp10', 'urban', 'wf_01',
       'wf_02', 'wf_03', 'wr_01', 'wr_02', 'wr_03', 'wr_04', 'wr_05', 'wr_06',
       'wr_07', 'wr_08', 'wr_09', 'wr_10', 'wr_11', 'wr_12', 'wr_13'],
      dtype='object')

In [15]:
# 연령대 -> 영어 -> 결혼 
zerodata = raw_data[raw_data['education']==0].copy()
nonzerodata = raw_data[raw_data['education'] != 0].copy()

In [16]:
groupnon = nonzerodata.groupby(['age_group','married','education']).size()
groupnon.head()

age_group  married  education
+70s       0        4            1
           1        1            2
                    2            6
                    3            4
                    4            6
dtype: int64

In [17]:
groupall = nonzerodata.groupby(['age_group','married']).size()
groupall

age_group  married
+70s       0              1
           1             18
           2            177
           3            103
10s        0             52
           1          17496
           2             93
           3             31
20s        0             22
           1          15245
           2           1873
           3            356
30s        0             16
           1           4361
           2           4042
           3           1203
40s        0             19
           1           1350
           2           3366
           3           1482
50s        0              4
           1            494
           2           1908
           3           1104
60s        0              5
           1            136
           2            865
           3            440
dtype: int64

In [18]:
groupper = groupnon / groupall
groupper # '+70s'

age_group  married  education
+70s       0        4            1.000000
           1        1            0.111111
                    2            0.333333
                    3            0.222222
                    4            0.333333
                                   ...   
60s        2        4            0.337572
           3        1            0.015909
                    2            0.313636
                    3            0.350000
                    4            0.320455
Length: 104, dtype: float64

In [19]:
groupper.index

MultiIndex([('+70s', 0, 4),
            ('+70s', 1, 1),
            ('+70s', 1, 2),
            ('+70s', 1, 3),
            ('+70s', 1, 4),
            ('+70s', 2, 1),
            ('+70s', 2, 2),
            ('+70s', 2, 3),
            ('+70s', 2, 4),
            ('+70s', 3, 1),
            ...
            ( '60s', 1, 3),
            ( '60s', 1, 4),
            ( '60s', 2, 1),
            ( '60s', 2, 2),
            ( '60s', 2, 3),
            ( '60s', 2, 4),
            ( '60s', 3, 1),
            ( '60s', 3, 2),
            ( '60s', 3, 3),
            ( '60s', 3, 4)],
           names=['age_group', 'married', 'education'], length=104)

In [20]:
num = len(zerodata)
na = []
for i in range(num):
    ll = list(zerodata[['age_group','married']].values[i])
    
    try:
        per = list(groupper[ll[0],ll[1]].values)
        number = list(groupper[ll[0],ll[1]].index)
    except Exception as e:
        print(i, e)
        per = [0.25,0.25,0.25,0.25]
        number = [1,2,3,4]
    na.append(np.random.choice(number,p=per))
    
zerodata['education'][zerodata['education'] == 0] = na
#zerodata[['age_group','engnat','married']].values

In [21]:
zerodata['education'].value_counts(),len(zerodata['education'])

(3    246
 2    237
 4    121
 1     46
 Name: education, dtype: int64,
 650)

In [22]:
datas = pd.concat([zerodata,nonzerodata])
datas.shape

(56912, 77)

In [23]:
datas['education'].value_counts()

2    21305
3    19271
4     9938
1     6398
Name: education, dtype: int64

In [24]:
datas

index  QaA  QaE  QbA   QbE  QcA   QcE  QdA   QdE  QeA  ...  wr_04  \
113      113  3.0  749  3.0   539  4.0   487  2.0   937  3.0  ...      1   
262      262  1.0  402  4.0  1117  4.0   676  1.0  1056  1.0  ...      1   
439      439  3.0  395  4.0  1700  1.0   664  2.0   562  3.0  ...      1   
537      537  1.0  456  5.0  1515  4.0  1312  2.0  3694  5.0  ...      1   
715      715  1.0  275  5.0  1307  4.0   861  1.0   615  5.0  ...      1   
...      ...  ...  ...  ...   ...  ...   ...  ...   ...  ...  ...    ...   
56910  56910  5.0  427  5.0  1066  5.0   588  1.0   560  2.0  ...      1   
56911  56911  1.0  314  5.0   554  5.0   230  1.0   956  2.0  ...      1   
56912  56912  1.0  627  2.0   799  1.0   739  2.0  1123  1.0  ...      1   
56913  56913  2.0  539  1.0  2090  2.0  4642  1.0   673  2.0  ...      1   
56914  56914  2.0  541  4.0   900  5.0   691  2.0  1951  1.0  ...      0   

       wr_05  wr_06  wr_07  wr_08  wr_09  wr_10  wr_11  wr_12  wr_13  
113        1      0      1      1      0      1      0      1      1  
262        1      0      1      1      0      1      0      1      1  
439        0      0      1      1      0      1      0      1      1  
537        1      0      1      1      0      1      1      1      1  
715        1      1      1      1      0      1      0      1      1  
...      ...    ...    ...    ...    ...    ...    ...    ...    ...  
56910      1      0      1      1      0      1      0      1      1  
56911      1      1      1      1      1      1      1      1      1  
56912      1      0      1      1      0      1      0      1      1  
56913      0      0      1      1      0      1      1      1      0  
56914      0      1      0      1      0      0      1      0      0  

[56912 rows x 77 columns]

In [28]:
datas[["QeA", "QfA", "QkA", "QqA", "QrA", "QaA", "QdA", "QgA", "QiA", "QnA"]]= (6 - datas[["QeA", "QfA", "QkA", "QqA", "QrA", "QaA", "QdA", "QgA", "QiA", "QnA"]])

In [36]:
datas['Mscore'] = datas.iloc[:,1::2].iloc[:,:20].sum(axis= 1)

In [43]:
datas.iloc[:,41:].corr()[datas.iloc[:,41:].corr()>0.4]['Mscore']

education          NaN
engnat             NaN
familysize         NaN
hand               NaN
married            NaN
tp01               NaN
tp02               NaN
tp03               NaN
tp04               NaN
tp05               NaN
tp06               NaN
tp07          0.447332
tp08               NaN
tp09               NaN
tp10               NaN
urban              NaN
wf_01              NaN
wf_02              NaN
wf_03              NaN
wr_01              NaN
wr_02              NaN
wr_03              NaN
wr_04              NaN
wr_05              NaN
wr_06              NaN
wr_07              NaN
wr_08              NaN
wr_09              NaN
wr_10              NaN
wr_11              NaN
wr_12              NaN
wr_13              NaN
Mscore        1.000000
Name: Mscore, dtype: float64